app/code/Magento/Theme/Observer/SetCookieThemeMessage.php

```diff
@@ -72,18 +80,31 @@ public function __construct(
      * Checks the result that controller actions must return. If result is not JSON type, then
      * sets 'mage-messages' cookie.
      *
-     * @param ResultInterface $subject
-     * @param ResultInterface $result
-     * @return ResultInterface
+     * @param Observer $observer
+     * @throws \Magento\Framework\Exception\InputException
+     * @throws \Magento\Framework\Stdlib\Cookie\CookieSizeLimitReachedException
+     * @throws \Magento\Framework\Stdlib\Cookie\FailureToSendException
      */
-    public function afterRenderResult(
-        ResultInterface $subject,
-        ResultInterface $result
-    ) {
-        if (!($subject instanceof Json)) {
+    public function execute(Observer $observer)
+    {
+        if ($this->isJsonResponse($observer) !== false) {
             $this->setCookie($this->getMessages());
+        };
+    }
+
+    /**
+     * Check If Response is Json
+     * @param Observer $observer
+     * @return bool
+     */
+    protected function isJsonResponse(Observer $observer)
+    {
+        try {
+            $this->serializer->unserialize($observer->getResponse()->getBody());
```

Comment:
I don't like this functionality, I am aware of that, but I don't know how I could validate that a $response is Json type